br/pkg/lightning/backend/local/local.go

```diff
@@ -808,50 +815,77 @@ func (local *local) WriteToTiKV(
 		}
 		clients = append(clients, wstream)
 		requests = append(requests, req)
+		storeIDs = append(storeIDs, peer.StoreId)
 	}
 
 	bytesBuf := local.bufferPool.NewBuffer()
 	defer bytesBuf.Destroy()
 	pairs := make([]*sst.Pair, 0, local.batchWriteKVPairs)
 	count := 0
 	size := int64(0)
+	totalSize := int64(0)
 	totalCount := int64(0)
-	firstLoop := true
 	// if region-split-size <= 96MiB, we bump the threshold a bit to avoid too many retry split
 	// because the range-properties is not 100% accurate
 	regionMaxSize := regionSplitSize
 	if regionSplitSize <= int64(config.SplitRegionSize) {
 		regionMaxSize = regionSplitSize * 4 / 3
 	}
+	// If there is a writeLimiter, we set a lower flush limit to make
+	// the speed of write more smooth.
+	flushLimit := int64(math.MaxInt64)
+	if local.writeLimiter != nil {
+		flushLimit = int64(local.writeLimiter.limit / 10)
+	}
+
+	flushKVs := func() error {
+		for i := range clients {
+			if local.writeLimiter != nil {
+				if err := local.writeLimiter.WaitN(ctx, storeIDs[i], int(size)); err != nil {
```

Comment:
Your analysis is correct. There is a `flushLimit`, which is not greater than the limit. So the `WaitN` size is usually less than the burst.

Reply:
In my test, the `size` is about 3M. In practice, there is no need to set such a small limit. Because to achieve this speed, we can use tidb-backend instead.

Reply:
> At that time where preparation finishes, 3+3+3 = 9 seconds has elapsed. But for each node, only 3000 bytes are sent. So the actual writing rate for each node is the sending size / total prepare time, which is 3000 / 9.

I thought of it again. I think this is as expected. Because kvs are sent to tikv node one by one for each region. The total send size is 9000, we need 9 seconds to send these data. It's unnecessary to wait for the rate limiter in parallel unless we actually send it data to tikv in parallel.